salt/modules/boto_s3.py

```diff
@@ -0,0 +1,176 @@
+# -*- coding: utf-8 -*-
+'''
+Connection module for Amazon S3 using boto3
+
+.. versionadded:: develop
+
+:configuration: This module accepts explicit AWS credentials but can also
+    utilize IAM roles assigned to the instance through Instance Profiles or
+    it can read them from the ~/.aws/credentials file or from these
+    environment variables: AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY.
+    Dynamic credentials are then automatically obtained from AWS API and no
+    further configuration is necessary. More information available at:
+
+    .. code-block:: text
+
+        http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/
+            iam-roles-for-amazon-ec2.html
+
+        http://boto3.readthedocs.io/en/latest/guide/
+            configuration.html#guide-configuration
+
+    If IAM roles are not used you need to specify them either in a pillar or
+    in the minion's config file:
+
+    .. code-block:: yaml
+
+        s3.keyid: GKTADJGHEIQSXMKKRBJ08H
+        s3.key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+
+    A region may also be specified in the configuration:
+
+    .. code-block:: yaml
+
+        s3.region: us-east-1
+
+    If a region is not specified, the default is us-east-1.
+
+    It's also possible to specify key, keyid and region via a profile, either
+    as a passed in dict, or as a string to pull from pillars or minion config:
+
+    .. code-block:: yaml
+
+        myprofile:
+            keyid: GKTADJGHEIQSXMKKRBJ08H
+            key: askdjghsdfjkghWupUjasdflkdfklgjsdfjajkghs
+            region: us-east-1
+
+:depends: boto3
+'''
+# keep lint from choking on _get_conn and _cache_id
+# pylint: disable=E0602
+
+# Import Python libs
+from __future__ import absolute_import
+import logging
+
+# Import Salt libs
+from salt.utils.versions import LooseVersion as _LooseVersion
+
+log = logging.getLogger(__name__)
+
+# pylint: disable=import-error
+try:
+    # pylint: disable=unused-import
+    import boto3
+    # pylint: enable=unused-import
+    import botocore
+    logging.getLogger('boto3').setLevel(logging.CRITICAL)
+    HAS_BOTO = True
+except ImportError:
+    HAS_BOTO = False
+# pylint: enable=import-error
+
+
+def __virtual__():
+    '''
+    Only load if boto libraries exist and if boto libraries are greater than
+    a given version.
+    '''
+    needed_boto3_version = '1.2.1'
+    msg = 'The boto_s3 module cannot be loaded: {0}.'
+    if not HAS_BOTO:
+        return (False, msg.format('boto3 libraries not found'))
+    if _LooseVersion(boto3.__version__) < _LooseVersion(needed_boto3_version):
+        submsg = 'boto3 library version {0} is required'.format(
+            needed_boto3_version,
+        )
+        return (False, msg.format(submsg))
+    return True
+
+
+def __init__(opts):  # pylint: disable=unused-argument
+    if HAS_BOTO:
+        __utils__['boto3.assign_funcs'](__name__, 's3')
+
+
+def get_object_metadata(
```

Comment:
I'm so sorry -- I got into a yellowjacket nest in the yard last Friday and have been largely useless since then.  This is my first chance to check back on this.

I'm a bit confused by this function.  You pass `**extra_args` (which is really just `**kwargs`) to the `head_object()` function without any munging or even validation, but you require them to be passed as `extra_args` rather than the seemingly more compact `**kwargs`.  Since you're doing no validation and no munging, I would request you change the function signature to 
```
def head_object(
    name,
    region=None,
    key=None,
    keyid=None,
    profile=None,
    **kwargs
)
```
Renaming to `head_object()` lets the salt exec function name match the underlying function it is directly calling, which makes documentation lookups for the end-user much simpler.  It also allows you to remove the `if extra_args is None:` stanza since kwargs will always be a dict.

I also suggest (request) that `Bucket` and `Key` be parsed from `name` ONLY in the event they are not explicitly passed.  Allowing arbitrary state names -- e.g. `name="Ensure my file in in the bucket"` is a standard salt vernacular.  Also, the behaviour of passing e.g. `Bucket='my_bucket', Key='path/to/file'` in your current code is undefined, since `Bucket` and `Key` will be defined twice in the `head_object()` call as it is written.

Reply:
The `extra_args` naming comes directly from the parameter name for upload_file, and the set of possible arguments is documented at http://boto3.readthedocs.io/en/latest/reference/customizations/s3.html?highlight=s3transfer#boto3.s3.transfer.S3Transfer.upload_file.
The bucket and key arguments are separate even in the boto3 API, so I'd prefer to keep them separate from extra args in the Salt API as well.
Salt provides the ability for arbitrary state IDs, but the state names are semantically significant. I agree that arbitrary IDs that are human readable are useful and we use them as well, but the name is meant to refer to the actual resource to be managed - see the examples in the 10-minute Salt tutorial (https://docs.saltstack.com/en/getstarted/config/functions.html).

Reply:
My question above was not about `upload_file()` but rather `head_object()`, which you use more-or-less directly in `get_object_metadat()` - `head_object()` does not have an `extra_args` (or an `ExtraArgs`) parameter, so as I mentioned I am confused as to why it was introduced here (or anywhere actually except in `upload_file()` where it actually is a function param (as `ExtraArgs`)  Again, my concern is about adding unneeded complexity (that is to say, `rakes`) for no particular reason, which will only serve to make maintainer's lives harder later, and yields no visible benefit now to justify it.